COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,458 @@
+
+##### `/ipfs/CID/*` `/ipfs/CID/P/A/T/H*`
+
+IPFS-Path-Prefix-Rule: Blocks any multihash-path combination starting with the
+the given path prefix. `/*` includes the empty path. Thus `/ipfs/CID/*`
+blocks the CID itself, and any paths. Examples:
+
+- `/ipfs/CID/*` : blocks CID (by multihash) and any path before resolving.
+- `/ipfs/CID/ab*`: blocks any path derived from the CID (multihash) and starting with "ab", including "ab"
+- `/ipfs/CID/ab/*`: equivalent to the above.
+
+Blocking layer recommendation: PathResolver + (BlockService if the CID itself is blocked too).
+
+##### `/ipns/IPNS`
+
+IPNS-rule: Blocks the given IPNS name before resolving. It does not block the CID that it
+resolves to.
+
+If the IPNS name is a domain name, it is blocked directy.
+
+If the IPNS name is a CIDv1 (libp2p-key) or b58-encoded-multihash (CIDV0),
+then the blocking affects the underlying Multihash.
+
+Blocking layer recommendation: NameSystem.
+
+##### `/ipns/IPNS/PATH`
+
+IPNS-Path-rule: Blocks specifically the IPNS path, before resolving. Equivalent to `/ipfs/CID/PATH`.
+
+Blocking layer recommendation: There is no good place to implement this rule
+as the NameSystem only handles IPNS names (without paths), and the
+path.Resolver only handles already-resolved Paths.
+
+##### `/ipns/NAME/*` `/ipns/NAME/PATH*`
+
+IPNS-Path-Prefix-Rule: Same as with the IPFS-Path-Prefix-Rule.
+
+Blocking layer recommendation:  There is no good place to implement this rule
+as the NameSystem only handles IPNS names (without paths), and the
+path.Resolver only handles already-resolved Paths.
+
+##### `/PATH` `/PATH/*` `/PATH*`
+
+Subpath-Rule: Block solely by looking at the subpath component of an IPFS path. Examples:
+
+- `/my/path`: blocks any item that tries to resolve `/my/path`, regardless of the CID used.
+- `/my/path*` and `/my/path/*`: blocks any paths that contain the prefix `/my/path`.
+
+Blocking layer recommendation: PathResolver.
+
+##### `//DOUBLE_HASH`
+
+Doublehash-Rule: Blocks using double-hashed item, which can be:
+
+- The sha256-hex-encoded hash of `CIDV1_BASE32/PATH`: this is the legacy
+  badbits block anchor format. It can only block by CID and not by
+  multihash. When no path present, the trailing slash must be kept
+  (`CIDV1_BASE32/`).
+- A b58-encoded multihash (a.k.a CIDV0), corresponding to the Sum() of:
+  - An IPNS-Path:
+    - `/ipns/IPNS` when the IPNS name is NOT a CID.
+    - The b58-encoded-multihash extracted from an IPNS key when the IPNS key
+      is a CID.
+  - An IPFS-Path: `b58-encoded-multihash/P/A/T/H` where the multihash is
+    extracted from the CID in `/ipfs/CID/P/A/T/H` (The multihash and the CID
+    are the same in the case of CIDV0). The `/P/A/T/H` component is optional
+    and should not have a trailing `/`.
+
+The latter form allows blocking by double-hash using any hashing function of
+choice. Implementations will have to hash requests using all the hashing functions
+used in the denylist, so we recommend sticking to one.
```

Comment:
Do we need this yet? Can we specify use of a multihash for self-describing future-proof goodness but require that only sha256 is valid for now to simplify things? For context: I'm looking at how i would add support for this spec to e-ipfs and w3s.link.

I acknowledge that we already have to accept providers can only do best effort enforcement of these lists given we already have to deal with path based rules which can't enforced at the BlockService level e.g in e-ipfs which is bitswap only... so i could just ignore entries that are hashed with a fn i don't support yet, but i'd prefer to be able to say "we support all block items that can be applied by the blockservice".

Reply:
If your denylist only includes one type of hashing function, then you only hash using that one to check the double-hashes. So it is up to who makes the denylist... I would totally discourage using multiple hashing functions for items in the same denylist.

But other than that, since they are multihashes there's no reason to support anything. In a way I'm deferring the discussion of "what is the best hashing function for this" until later, because sha256 is probably not it. Mostly you just want something very very fast, and doesn't necessarily need to be a cryptographic one as it just meant to obfuscate.

Reply:
as above, the tradeoffs are expressiveness of list-sharing for compute-obligation for conforming processors of the lists.  here too deciding layering before committing to a degree of expressiveness (or the core/extension, required/optional corrolary) might be preferable to deciding this in advance of commitments to implement and conform at scale.

Reply:
I am fine prominently recommending a hashing function. Which one should it be though?

What is:

* A cryptographic hash function (we need this in the end as pre-image and collision protection are important).
* Well supported in all languages/platforms
* Very fast

Is it blake3? sha256? Are we ok with md5 (sure there are collisions but need to be carefully crafted). ? I'll ask around.

Reply:
Would you consider defining it as _only_ sha2-256 for the first version of the spec? That restriction could be loosened later if folks need it.

Reply:
I would like more input. I think sha2-256 may impose larger perf impact than necessary.

Reply:
Here is an idea/compromise: we explicitly define the hash functions that should be used in the YAML header under `double-hash` key. If the key is not present, we assume SHA256 and clients only use that (backward-compatible). 

This way, we future-proof the format to switch to better functions as the time goes by, but we don't force everyone to waste resources on hashing with every function if the list only uses one.

Example: if we want to limit users to Blake3 `double-hash: 0x1e` or  allow SHA256 and Blake3: `double-hash: 0x1e, 0x12`

Reply:
Clients don't need to hash with every function. They only hash with the functions that the list uses, and they know which ones are those because of the multihash-prefix, so as you parse the list you know which ones to double-hash with (if any).

The future-proofness is ensured by using multihashes already.

The question is whether we require those multihashes to be sha256 only or something else.

Reply:
For pragmatic reasons, our options are probably limited to either sha2-256 or blake3 (ipfs impls. will have both).
Between sha2 and blake3, the latter should make a faster default (faster than md5 too). Maybe benchmark?

ps. The purpose behind the explicit list of hashes in the YAML header was to protect the node operator from silently exploding the list of hashes to check when the intention is to only use one (or two) functions.  But maybe it is not a real DoS risk, these lists will come from somehow trusted sources.
If we decide that we only use a single hash function per list, it still feels like something that should be explicitly defined in YAML header, and not hard-coded in the spec (future-proofing without having to change the specs).

Reply:
yes, i  strongly agree that the header yaml should define the hash fn, and i am asking that the spec, in it's initial wording, require only sha-256 be used.

we can even add a note in the spec that folks should raise a PR here if that is causing them real perf problems.

adding support for more hash functions would have to take into account that there is no baseline requirement of what hash functions an ipfs impl must support. web3-storage does not verify blake3 today. surprising but true https://github.com/web3-storage/car-block-validator/blob/cb121b780c86724c0266f797b5a4badffad77140/src/index.js#L1-L17

Reply:
You don't need to list hash functions in the header. As you parse the list, you see the multihash function used in the multihashes that you encounter and that is the list of hashing functions you have double-hash with. Putting the list at the top is redundant.

I think the issue here is a matter of compatibility, and sha256 being the best thing that is nicely and universally supported everywhere. I'm ok with the spec requiring sha256 for double-hashing, otherwise leave things open as they are. If people use other hashes (which remains possible), they would be breaking the spec, but that's up to them.

Reply:
I actually disagree that listing which hash functions are used in a list is necessary and always _redundant_, because I am focused on a use-case where more lists from more unfamiliar authorities get added and subtracted more often, making the list itself worth annotating.  If managing and federating many denylists is being optimized for, all kinds of (optional!) frontmatter fields could be quite helpful in deciding _which ones_ to federate/follow before parsing them, particularly for software that builds on CIDs selectively (which might need to know in advance if a denylist includes entries it cannot, or does not need to, parse).

In any case, I would argue that an optional frontmatter field for declaring which hash functions were/are used to populate a list is good, actually.  I don't have strong opinions, but some basic questions, in case enough people agree that a section is worth adding to the IPIP itself:
1. should the values be `name` entries from the [`multicodec` table](https://github.com/multiformats/multicodec/blob/master/table.csv), or the actual tags?
2. what should a consumer do/think/feel if an entry uses a tag not announced up top?
3. should list hosters ADD entries to the frontmatter declaration over time if they add functions over time, or create a new list at a new URL when these "env var"s change? Considering many subscribers have to unload and reload the list for any changes except EOF appends, this could complicate the spec a lot.

Reply:
1. name for sanity
2. With or without hash-function annotations at the top, you're in the same situation if you find a hash that you don't support while parsing. Annotation at the top would only help to reject a list altogether before parsing it. But why would you do that and not try to work with the entries that you can support in any case? Anyways, you can use any optional hints you want. We can standardize an optional hint name for this as long as it's optional.
3. I think you're finding the subtle issues. "whoops now my list includes also blake3 double-hashes" is bad to communicate in any way. "List A is done and you can follow up again on list B" is also not considered. It would be easier if we specify what the recommended behavior is when finding an unknown hashing function.

Also, we are talking a lot about hashing functions pertaining to the double-hash entries that are enconded as multihashes. You could also have normal `/ipfs/<cid>` entries, where the CID itself is not something that your underlying IPFS implementation understands: maybe you don't know the ipld codec, maybe the hash-function is unknown. In this cases, it would not be the "compact denylist parser component/blocker" the one that needs to be aware, but the underlying "ipfs" implementation. etc. but some implications might be similar.

So my practical approach is, 1) everything goes 2) try to parse what you can 3) strong recommendation or obligation to only use sha256 for double-hashing, everything else is non-spec-compliant (for now).